### ISSUE TITLE
Button Block on hover: contrast ratio issue when setting custom colors

### ISSUE THREAD

**Issue by briceduclos** [WordPress/twentytwentyone#612]
**Describe the bug**

While the hover effect (fill to outline and vice versa) works for the default buttons, as soon as the user tries to customize the colors of the button block, the hover effect makes it difficult to have a button with a good contrast ratio.

### Fill Button Examples

- Button 1: custom background-color, default text color. On hover, the text becomes black, contrast ratio too low.
- Button 2: custom background-color, text color white (from the color palette). On hover, the text still becomes black, contrast ratio too low.
- Button 3: custom background-color, custom text color. The hover state is the same as the “default” state.

"Default" state
![button](https://user-images.githubusercontent.com/56188815/96571703-35318c00-12cc-11eb-8e58-7067755a1b51.PNG)
Hover state
![button-hover](https://user-images.githubusercontent.com/56188815/96571705-35318c00-12cc-11eb-9254-f76f84cfe54a.PNG)

### Outline Buttons Examples

- Button 1: text color gray (from the color palette): works inside the editor, doesn’t work on the front. All the colors from the palette don’t work on the front, they are overriden.
- Button 2: custom text color. On hover, the background becomes black, contrast ratio too low.

Default state
![button-outline](https://user-images.githubusercontent.com/56188815/96571700-34005f00-12cc-11eb-8be0-7b5aa8352326.PNG)
Hover state
![button-outline-hover](https://user-images.githubusercontent.com/56188815/96571701-3498f580-12cc-11eb-9a9d-58e450a6d263.PNG)

A solution might be to keep the effect for the default buttons, but not apply the hover effect when the user specifies a text color or a background color.

**To reproduce**

Here's the code:

```
<!-- wp:buttons -->
<div class="wp-block-buttons"><!-- wp:button {"style":{"color":{"background":"#007c89"}},"className":"is-style-fill"} -->
<div class="wp-block-button is-style-fill"><a class="wp-block-button__link has-background" href="#" style="background-color:#007c89">Button 1</a></div>
<!-- /wp:button -->

<!-- wp:button {"style":{"color":{"background":"#0070c9"}},"textColor":"white","className":"is-style-fill"} -->
<div class="wp-block-button is-style-fill"><a class="wp-block-button__link has-white-color has-text-color has-background" href="#" style="background-color:#0070c9">Button 2</a></div>
<!-- /wp:button -->

<!-- wp:button {"style":{"color":{"text":"#eeeeee","background":"#0070c9"}},"className":"is-style-fill"} -->
<div class="wp-block-button is-style-fill"><a class="wp-block-button__link has-text-color has-background" style="background-color:#0070c9;color:#eeeeee">Button 3</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->

<!-- wp:buttons -->
<div class="wp-block-buttons"><!-- wp:button {"textColor":"gray","className":"is-style-outline"} -->
<div class="wp-block-button is-style-outline"><a class="wp-block-button__link has-gray-color has-text-color" href="#">Button 1</a></div>
<!-- /wp:button -->

<!-- wp:button {"style":{"color":{"text":"#0070c9"}},"className":"is-style-outline"} -->
<div class="wp-block-button is-style-outline"><a class="wp-block-button__link has-text-color" href="#" style="color:#0070c9">Button 2</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->
```

**Comment by carolinan** [comment]
@briceduclos Thank you for a great issue description. We knew there were color issues left but it helps to have them narrowed down this way.
